src/examples/librpmem/fibonacci/fibonacci.c

```diff
@@ -0,0 +1,403 @@
+		return NULL;
+	}
+
+verify:
+	/* verify signature */
+	if (strcmp(pool_attr.signature, POOL_SIGNATURE) != 0) {
+		pool_attr.signature[RPMEM_POOL_HDR_SIG_LEN - 1] = '\0';
+		fprintf(stderr, "invalid signature: %s != %s\n",
+				pool_attr.signature, POOL_SIGNATURE);
+		goto err;
+	}
+
+	return rpp;
+
+err:
+	/* close the remote pool */
+	if (rpmem_close(rpp)) {
+		fprintf(stderr, "rpmem_close: %s\n", rpmem_errormsg());
+		exit(1);
+	}
+
+	return NULL;
+}
+
+static int
+remote_write(RPMEMpool *rpp)
+{
+	printf("storing Fibonacci numbers on the target...\n");
+	if (rpmem_persist(rpp, DATA_OFF, DATA_SIZE, 0, 0)) {
+		printf("store failed: %s\n", rpmem_errormsg());
+		return 1;
+	}
+	return 0;
+}
+
+static int
+remote_read(RPMEMpool *rpp, void *buff)
+{
+	printf("restore Fibonacci numbers from the target...\n");
+	if (rpmem_read(rpp, buff, DATA_OFF, DATA_SIZE, 0)) {
+		printf("restore failed: %s\n", rpmem_errormsg());
+		return 1;
+	}
+	return 0;
+}
+
+static int
+fibo_checksum(struct fibo_t *fibo)
+{
+	return fibo->signature[0] + fibo->fn + fibo->fn1;
+}
+
+static void
+fibo_init(struct fibo_t *fibo)
+{
+	printf("initializing...\n");
+	memset(fibo, 0, DATA_SIZE);
+	strncpy(fibo->signature, POOL_SIGNATURE, FIBO_SIG_LEN);
+	fibo->n = 0;
+	fibo->fn = 0;
+	fibo->fn1 = 1;
+	fibo->checksum = fibo_checksum(fibo);
+}
+
+static int
+fibo_is_valid(struct fibo_t *fibo)
+{
+	if (strncmp(fibo->signature, POOL_SIGNATURE, FIBO_SIG_LEN) != 0)
+		return 0;
+
+	return fibo->checksum == fibo_checksum(fibo);
+}
+
+static int
+fibo_is_zeroed(struct fibo_t *fibo)
+{
+	char *data = (char *)fibo;
+	for (size_t i = 0; i < DATA_SIZE; ++i)
+		if (data[i] != 0)
+			return 0;
+
+	return 1;
+}
+
+/*
+ * fibo_validate -- validate local and remote copies of the F sequence
+ */
+static struct fibo_t *
+fibo_validate(struct fibo_t *fibo, struct fibo_t *fibo_r, unsigned *state)
+{
+	struct fibo_t *valid = NULL;
+
+	if (fibo_is_valid(fibo))
+		valid = fibo;
+	else if (!fibo_is_zeroed(fibo)) {
+		fprintf(stderr, "broken local memory pool!\n");
+		*state |= BROKEN_LOCAL;
+	}
+
+	if (fibo_is_valid(fibo_r)) {
+		if (!valid)
+			valid = fibo_r;
+		else if (fibo_r->n > valid->n)
+			valid = fibo_r;
+	} else if (!fibo_is_zeroed(fibo_r)) {
+		fprintf(stderr, "broken remote memory pool!\n");
+		*state |= BROKEN_REMOTE;
+	}
+
+	if (!valid)
+		printf("no valid Fibonacci numbers found.\n");
+
+	return valid;
+}
+
+/*
+ * fibo_cleanup -- based on validation outcome cleanup copies and initialize
+ * if required
+ */
+static void
+fibo_cleanup(RPMEMpool *rpp, void *pool, struct fibo_t *fibo_r,
+		int *initialized)
+{
+	struct fibo_t *fibo = GET_FIBO(pool);
+	struct fibo_t temp;
+	unsigned state = 0;
+	int ret;
+
+	struct fibo_t *valid = fibo_validate(fibo, fibo_r, &state);
+
+	/* store valid fibonacci data in temp */
+	if (valid)
+		memcpy(&temp, valid, sizeof(temp));
+
+	/* local cleanup */
+	if (state & (BROKEN_LOCAL | BROKEN_REMOTE))
+		memset(pool, 0, POOL_SIZE);
```

Comment:
Why do you zero everything when one replica is corrupted?

Reply:
- if the local pool is corrupted it has to be zeroed
- if the remote pool is corrupted the local pool has to be zeroed and its content has to be replicated to the remote. Afterwards, previous local pool content is restored for further processing.

Reply:
1) ok
2) but that's against the whole point of replication - if a remote pool is corrupted its content should be restored from the local one

Reply:
You are right. When the generator state was stored on heap and both replicas are zeored the generator state is not persistent. I think it is now fixed.